### PR TITLE
Decouple asFlow from batchSize and move it to buffer instead, promote…

### PR DESCRIPTION
**Pull request by qwwdfsad**
… it to experimental

### CODE REVIEW

File: reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt [hunk 1]
```diff
@@ -7,66 +7,111 @@ package kotlinx.coroutines.reactive.flow
...
-@JvmOverloads // For nice Java API
 @JvmName("from")
-public fun <T : Any> Publisher<T>.asFlow(batchSize: Int = 1): Flow<T> =
-    PublisherAsFlow(this, batchSize)
+@Deprecated(
+    message = "batchSize parameter is deprecated, use .buffer() instead to control the backpressure",
+    level = DeprecationLevel.ERROR,
+    replaceWith = ReplaceWith("asFlow().buffer(batchSize)", imports = ["kotlinx.coroutines.flow.*"])
+)
+public fun <T : Any> Publisher<T>.asFlow(batchSize: Int): Flow<T> = asFlow().buffer(batchSize)
```
Comment (qwwdfsad) [inline reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt]: It breaks binary compatibility around `asFlow$default`, but it should be okay for preview feature

Comment (elizarov) [reply reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt]: Look fine.

**Comment by elizarov** [comment]
I was not looking to optimize this case, but to have a single `Subscriber` implementation. They are hard to get right, so having just one such implementation is easier it terms of testing and future maintenance. However, `Subscriber` implementation that `openSubscription` has (which I've used) it definitely more complex. Moreover, it has different behavior. It uses "batch size" only for initial call to `request` but then just makes constant `request` calls just for one element in the future. That is less efficient that the original `Subscriber` implementation in this PR and may worth a look at.

**Comment by elizarov** [comment]
Double checked... that is not the case... `SubscriptionChannel` has similar batch-requesting logic (requests full batch when it is over), but it has weird behavior with respect to batch. I'll looking at how valid it is.

**Comment by elizarov** [comment]
Anyway, this solution would not work with `publisher.asFlow().conflate()`. Needs to be fixed.

**Comment by elizarov** [comment]
Ok. Having studied `openSubscription` in detail I've reverted to mostly original implementation of the subscriber for `publisher.asFlow()` because it is slightly more efficient. I've also fixed support for `conflate()` and properly integration with reactive contract for "requests everything" (shall be Long.MAX_VALUE!). Added tests for various case.

**Comment by elizarov** [comment]
@qwwdfsad Please take a look again.

File: reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt [hunk 2]
```diff
@@ -7,66 +7,124 @@ package kotlinx.coroutines.reactive.flow
...
         try {
-            var consumed = 0
-            for (i in channel) {
-                collector.emit(i)
-                if (++consumed == batchSize) {
-                    consumed = 0
-                    subscriber.subscription.request(batchSize.toLong())
+            var consumed = 0L
+            while (true) {
+                val value = subscriber.takeNextOrNull() ?: break
```
Comment (qwwdfsad) [inline reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt]: this instead of the iterator to save an extra object?

Comment (elizarov) [reply reactive/kotlinx-coroutines-reactive/src/flow/PublisherAsFlow.kt]: Yes

**Pull request merged**